fix(cli): Add fallback module resolver for `expo` and `expo-router` dependencies

kitten opened the pull request:
# Why

This adds a fallback module resolution strategy to our `withMetroMultiPlatform` resolver chain, which is intended to provide fallbacks when a failed resolution occurs via normal Node module resolution.

The dependencies and optional peer dependencies in `expo` and `expo-router` are special. Several package manager issues, quirks, or the isolated dependency strategy can cause their dependencies to either become inaccessible or not be hoisted. There are several cases in which this may happen.

> [!NOTE]
> Unlike the [sticky resolution strategy](https://github.com/expo/expo/pull/31750). This does not prevent duplicate dependencies, since it serves as a fallback. While sticky resolution is useful for **native module** dependencies (since only one may be built into an app at a time), it's too aggressive for missing module resolutions that occur due to package manager bugs, invalid hoisting trees, and isolated dependencies for non-native modules.

<details>
<summary><strong>Reminder Note:</strong> What are isolated dependencies?</summary>

For a reminder re. isolated dependencies. This is an installation strategy used by `pnpm`. In this installation structure it enforces that hoisted dependencies (and direct dependencies) are isolated to the actual dependencies that these modules specify.

In this strategy, modules that specify dependencies are linked to a central `node_modules/.pnpm` folder. In this folder, each dependency is placed in its own sub folder and its dependencies are recursively linked to other folders of this style.

This effectively means that no dependencies that aren't specified are accessible, avoiding "implicit transitive" dependencies, which can otherwise lead to bugs when hoisting differs due to conflicts.

</details>

### Isolated dependencies with transformations

Output ASTs are transformed using Babel. This may leave `@babel/runtime` helpers around. `@babel/runtime` is generally an entrypoint into several output helpers that Babel places into output bundles.

However, due to isolated dependencies, without this fallback resolvers, users will have to install `@babel/runtime` directly and rely on deduplication. This doesn't apply with hoisting, so it's unexpected and a repository can fail when the package manager is changed over to `pnpm` without the hoisted linker.

### Unhoisted dependencies

If any dependency of `expo` or `expo-router` isn't hoisted, it becomes inaccessible to the user. While this is expected, if this is either the result of a transformation or the result of transitive dependency rules, the module won't be resolved.

This occurred for `expo-asset/tools/hashAssetFiles` (which will be removed) in non-isolated dependencies cases too due to invalid hoisting. This is an intermittent error, where the dependency tree isn't technically incorrect, but hoisting is skipped for a dependency.

### Isolated/unhoisted peer dependencies that are auto-installed

Due to a quirk in modern package manager behaviour. Peer Dependencies may be auto-installed but not hoisted in some cases. This depends on the package managers' configuration.

We have several optional peer dependencies in `expo` and `expo-router` that may become inaccessible from the project root due to this problem, and if they're imported in the project, rather than in their own module code, an error will occur. This is usually user error, but also unexpected and hard to spot.

# How

The resolver created by `createExpoFallbackResolver` is added as the last one in our chain of custom resolvers. Importantly, it's placed after `requestPostRewrites` (previously unnamed), the prior last resolver, which always returns a resolution (or throws).
Hence, the fallback resolver is only used when normal Node module resolution fails.

`createExpoFallbackResolver` accepts a list of `originModuleNames` (we pass `'expo'` and `'expo-router'`). It then:
- resolves the root of these origin modules via a regular `require.resolve` call (meaning, all origin modules must be accessible via `@expo/cli` and/or the project, which is expected for `expo` and optionally `expo-router`)
- loads these modules' `package.json` and creates a regex for modules that are either their dependencies or optional peer dependencies

The resolver it creates then activates when we cannot resolve a module prior and:
- checks that the `moduleName` is a module specifier
- checks the `moduleName` against the regexes created above
- if a match occurs, it resolves the `moduleName` from the origin module's installation folder instead

## Set of changes

- Create type for `StrictResolver` and `StrictResolverFactory`
- Give names to resolvers passed to `withMetroResolvers` and add their names to the `debug` error message
- Add `createExpoFallbackResolver`

# Test Plan

- Added resolution tests to `withMetroMultiPlatform`'s tests

**Reproduction:**
- Clone https://github.com/kitten/expo-pnpm-monorepo-reproduction
  - _Note:_ The reproduction is a regular pnpm workspace with an Expo template app added to it as a workspace package. **Note that there's no `.npmrc` containing a hoisted dependency setting.** This triggers the related bug due to isolated dependencies.
- Run `pnpm install`
- Go to `apps/native-app/`
- Run `pnpm start --clear` and open the app (iOS or Android)
  - _Note:_ This will reproduce the related build error and will show `Unable to resolve "@babel/runtime/helpers/interopRequireDefault"...`

**Verification:**
- You can verify that `@babel/runtime` is not a direct dependency of the app (either through `pnpm list @babel/runtime` or by checking that `node_modules/@babel/runtime` does not exist)
- Adding `pnpm add @babel/runtime` would resolve this, since it manually provides a resolution (If you do this, make sure to reset this before testing the fix)

**Testing this fix:**
- Run `git checkout pr-34286 `
  - _Note:_ This branch contains this PR's changes applied as a pnpm patch (patch-package change)
- Run `pnpm install`
- Run `pnpm start --clear`

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

kitten committed:
Add strict resolver types

kitten committed:
debug: Add resolver names for withMetroResolvers throw debug message

kitten committed:
Implement fallback module resolver

kitten committed:
Add comments to `createExpoFallbackResolver`

kitten committed:
Add usage comment to withMetroMultiPlatform

byCedric commented on packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts:
```diff
@@ -0,0 +1,135 @@
+
+  return function requestFallbackModule(immutableContext, moduleName, platform) {
+    // Early return if `moduleName` cannot be a module specifier
+    if (moduleName[0] === '.' || moduleName[0] === '/') {
+      return null;
+    }
```
Does this also need to account for alias tokens? E.g. `@/module` or `~/module`?

kitten replied:
not necessarily. The module resolver (almost always) triggers when module resolution has otherwise failed. But this is just to avoid the regex checks as often as possible for imports that can never be module specifiers. It's ok if we miss some (and things like the tsconfig resolver come first anyway), since the regexes are what match the module specifiers we want to target

kitten committed:
Resolve origin modules through metro

This aligns it with the sticky resolver implementation and should
protect us from symlinking cases. Without this, if `@expo/cli` is
external to the current project it's used in, we'd resolve fallbacks
from the external `@expo/cli` dependency rather than from the local
project.

kitten committed:
Replace require with fs read in createExpoFallbackResolver

kitten committed:
Add initial fallback module resolution test

kitten committed:
Catch resolution error of `resolve` function for origin module loading

kitten committed:
Add additional fallback resolution tests

kitten committed:
Add test for non-matching dependencies

kitten committed:
Add CHANGELOG entry

kitten committed:
Fix type issue on filePath

kitten commented on packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts:
```diff
@@ -376,7 +383,11 @@ export function withExtendedResolver(
   const metroConfigWithCustomResolver = withMetroResolvers(config, [
     // Mock out production react imports in development.
-    (context: ResolutionContext, moduleName: string, platform: string | null) => {
+    function requestDevMockProdReact(
```
I gave the functions here names, since it was useful while debugging when I was adding extra logs and observing whether a prior one threw. It's unrelated to this PR, but could be nice to keep. Revertable via this commit: https://github.com/expo/expo/pull/34286/commits/8e722f26f50e1a89555154fc175be4624544789e

EvanBacon reviewed (approved):
lfg

EvanBacon commented on packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts:
```diff
@@ -0,0 +1,160 @@
+  let filePath: string | undefined;
+  let packageMeta: PackageMeta;
+  try {
+    const resolution = resolve(`${originModuleName}/package.json`);
```
There are some built-in methods on the Resolution context that may resolve better/faster `context.getPackage(pkgFile)`

kitten replied:
done!

EvanBacon commented on packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts:
```diff
@@ -0,0 +1,160 @@
+
+  return function requestFallbackModule(immutableContext, moduleName, platform) {
+    // Early return if `moduleName` cannot be a module specifier
+    if (moduleName[0] === '.' || moduleName[0] === '/') {
```
Are there not more rules here? Like `~`?

EvanBacon replied:
Do we need to account for windows here? Also is this an application-only feature, e.g. should we skip when the origin is in a node_module https://github.com/expo/expo/blob/48b2c22ebaeac472f18f403d191ae7408895232a/packages/%40expo/cli/src/utils/tsconfig/resolveWithTsConfigPaths.ts#L24-L35

byCedric replied:
@EvanBacon The module import specifiers in Metro are always resolved as POSIX formatted paths (with forward slashes). Metro does not support importing something like `import '.\path\to\file'`, meaning we don't have to account for that here.

Only when actually resolving, and using the paths to infer other things, we need to take Windows vs normal paths into account.

kitten committed:
Add comments explaining fallback resolver modifications

kitten committed:
Use Metro's context.getPackage to resolve package.json

kitten committed:
Improve module specifier check
